### PR TITLE
feat: support OutputChunk#dynamicImports and RenderedChunk#dynamicImp…

### PR DESCRIPTION
**Pull request by underfin**
…orts

### CODE REVIEW

**Review by hyf0 — APPROVED** [review]

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs [hunk 1]
```diff
@@ -234,47 +236,29 @@ impl<'a> GenerateStage<'a> {
...
-            .map(|id| {
-              self.link_output.module_table.normal_modules[*id].resource_id.expect_file().as_str()
-            })
-            .collect::<Vec<_>>();
-          resource_ids.sort_unstable();
-          resource_ids
-        });
-        cross_chunk_imports
-      })
-      .collect::<Vec<_>>();
```
Comment (underfin) [inline crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: The `cross_chunk_dynamic_imports` is collected by chunk modules `import_records`, the chunk modules is sorted, so the order is stable.
The `cross_chunk_imports` is from `chunk_meta_imports_vec`, it is collected by chunk modules `referenced_symbols`, the order is also stable.

Comment (hyf0) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: Didn't notice this part. Could you keep the original sorting logic. It have better stability on calculating chunk hash. https://github.com/rolldown/rolldown/pull/1002/files#r1583073224

---

Just notice you change it from hashset to vec.

**Review by hyf0 — CHANGES_REQUESTED** [review]

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs [hunk 2]
```diff
@@ -9,14 +9,14 @@ use rolldown_common::{
...
-use rolldown_utils::rayon::{IntoParallelIterator, ParallelBridge, ParallelIterator};
+use rolldown_utils::rayon::{ParallelBridge, ParallelIterator};
 use rustc_hash::{FxHashMap, FxHashSet};
 
 type ChunkMetaImports = IndexVec<ChunkId, FxHashSet<SymbolRef>>;
 type ChunkMetaImportsForExternalModules =
   IndexVec<ChunkId, FxHashMap<ExternalModuleId, Vec<NamedImport>>>;
 type ChunkMetaExports = IndexVec<ChunkId, FxHashSet<SymbolRef>>;
-type IndexCrossChunkImports = IndexVec<ChunkId, FxHashSet<ChunkId>>;
+type IndexCrossChunkImports = IndexVec<ChunkId, Vec<ChunkId>>;
```
Comment (hyf0) [inline crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: How `Vec<ChunkId>` avoid duplicate imports to the same chunk?

Comment (underfin) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: Yeah. Here should be `IndexSet`. The `chunk_meta_imports` is also has same problem.

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs [hunk 3]
```diff
@@ -234,47 +236,29 @@ impl<'a> GenerateStage<'a> {
...
         ),
       )
       .par_bridge()
       .for_each(
         |(
           chunk,
-          (imports_from_other_chunks, (imports_from_external_modules, cross_chunk_imports)),
+          (
+            imports_from_other_chunks,
+            (imports_from_external_modules, (cross_chunk_imports, cross_chunk_dynamic_imports)),
```
Comment (hyf0) [inline crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs:281]: Could you keep the `index` prefix above. The above `cross_chunk_imports: IndexVec<ChunkId, Vec<ChunkId>>` with this `cross_chunk_imports: Vec<ChunkId>>` could be really confusing,

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs (hunk 2 above)
Comment (hyf0) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: I don't think so. If the order relies on the order of traversing symbols, we might break the order due to adjust the generated runtime symbols by ourselves.

Depending the order calculated by `sorting with modules` is much reliable than this.

Comment (underfin) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: Ok. Let me revert it.

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs (hunk 3 above)
Comment (underfin) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs:281]: Updated

**Comment by underfin** [comment]
Here i revert the `index_sorted_cross_chunk_imports` back, but the order  is not correct, it seems to be the  execute order of chunk first module .

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs (hunk 2 above)
Comment (hyf0) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: Saw you were saying revert it back in https://github.com/rolldown/rolldown/pull/1002#issuecomment-2082796466. Did you mean rollup ensure the order of `imports` is the execution order of chunk?

Comment (underfin) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: Yeah. The rollup using js `Set/Map`, the Set and Map have the  insertion order, the insertion order is stabled.

Comment (hyf0) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: >  The rollup using js Set/Map, the Set and Map have the insertion order, the insertion order is stabled.

I don't get it. How this is the reason you are using `Vec<ChunkId>` now?

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs [hunk 4]
```diff
@@ -1,22 +1,26 @@
...
+use rolldown_utils::rayon::IntoParallelIterator;
+use rolldown_utils::rayon::{ParallelBridge, ParallelIterator};
+use rustc_hash::{FxHashMap, FxHashSet, FxHasher};
 
 type ChunkMetaImports = IndexVec<ChunkId, FxHashSet<SymbolRef>>;
 type ChunkMetaImportsForExternalModules =
   IndexVec<ChunkId, FxHashMap<ExternalModuleId, Vec<NamedImport>>>;
 type ChunkMetaExports = IndexVec<ChunkId, FxHashSet<SymbolRef>>;
-type IndexCrossChunkImports = IndexVec<ChunkId, FxHashSet<ChunkId>>;
+type IndexCrossChunkImports = IndexVec<ChunkId, Vec<ChunkId>>;
```
Comment (hyf0) [inline crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: Let me try to make things clear again:

Do not change `IndexCrossChunkImports` from `IndexVec<ChunkId, FxHashSet<ChunkId>>` to `IndexVec<ChunkId, Vec<ChunkId>>`.

- `Vec<ChunkId>` may contains duplicated `ChunkId`
- Don't use `IndexSet` on `IndexCrossChunkImports`. https://github.com/rolldown/rolldown/pull/1002/files#r1583073224

**Review by hyf0 — CHANGES_REQUESTED** [review]

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs (hunk 2 above)
Comment (underfin) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: I missing to revert it.

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs (hunk 4 above)
Comment (underfin) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: I missing revert it.

**Comment by underfin** [comment]
I revert it back, but the code need to rewrite with stable order.
Look like change `FxHashSet` to `IndexSet` at `type ChunkMetaImports = IndexVec<ChunkId, FxHashSet<SymbolRef>>;` `type IndexCrossChunkImports = IndexVec<ChunkId, FxHashSet<ChunkId>>;` is expected. 
The `imports_from_other_chunks_vec` is also need to using `IndexMap` instead of `HashMap`

**Comment by hyf0** [comment]
> I revert it back, but the code need to rewrite with stable order.

No need unless we relies on this kind of [order](https://github.com/rolldown/rolldown/pull/1002#issuecomment-2082796466) and we don't.

Relying on execution order will also make hash changed frequently than using order from `sorting module resourcesId`.

**Review by hyf0 — APPROVED** [review]

**Comment by underfin** [comment]
> > I revert it back, but the code need to rewrite with stable order.
> 
> No need unless we relies on this kind of [order](https://github.com/rolldown/rolldown/pull/1002#issuecomment-2082796466) and we don't.
> 
> Relying on execution order will also make hash changed frequently than using order from `sorting module resourcesId`.

It will cause the chunk order is not correct. let me give a example for you later.

**Comment by hyf0** [comment]
> > > I revert it back, but the code need to rewrite with stable order.
> > 
> > 
> > No need unless we relies on this kind of [order](https://github.com/rolldown/rolldown/pull/1002#issuecomment-2082796466) and we don't.
> > Relying on execution order will also make hash changed frequently than using order from `sorting module resourcesId`.
> 
> It will cause the chunk order is not correct. let me give a example for you later.

It's known by me and that's why https://github.com/rolldown/rolldown/issues/122 remains open. I'm still solving it with considering the stability of hash and correctness of execution order rather than only one aspect. Leave it to me.

**Pull request merged**